src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs

```diff
@@ -0,0 +1,347 @@
+                _patternMatcher = new PatternMatcher(pattern, allowFuzzyMatching: true);
+            }
+
+            public void Dispose()
+            {
+                _patternMatcher.Dispose();
+            }
+
+            internal async Task<ImmutableArray<INavigateToSearchResult>> DoAsync()
+            {
+                // Search all documents in parallel.  Each search gets a dedicated 'temp array' 
+                // that it can place its results into without having to worry about locking.  
+                // Once all the tasks are complete, we'll aggregate all the 'temp arrays' into
+                // the final result.
+                var documentSearchTasks = ArrayBuilder<Task>.GetInstance();
+                var tempArrays = ArrayBuilder<ArrayBuilder<INavigateToSearchResult>>.GetInstance();
+
+                try
+                {
+                    // Spawn off a task per document to search.
+                    SpawnDocumentSearchTasks(documentSearchTasks, tempArrays);
+
+                    // Wait for all the tasks to finish.
+                    await Task.WhenAll(documentSearchTasks).ConfigureAwait(false);
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    // Collect all the results into the final result.
+                    var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+                    foreach (var tempArray in tempArrays)
+                    {
+                        result.AddRange(tempArray);
+                    }
+
+                    return result.ToImmutableAndFree();
+                }
+                finally
+                {
+                    foreach (var tempArray in tempArrays)
+                    {
+                        tempArray.Free();
+                    }
+
+                    tempArrays.Free();
+                    documentSearchTasks.Free();
+                }
+            }
+
+            private void SpawnDocumentSearchTasks(ArrayBuilder<Task> tasks, ArrayBuilder<ArrayBuilder<INavigateToSearchResult>> tempArrays)
+            {
+                foreach (var document in _project.Documents)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    if (_searchDocument != null && document != _searchDocument)
+                    {
+                        continue;
+                    }
+
+                    var tempArray = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+                    tempArrays.Add(tempArray);
+
+                    // Kick off a fresh task so searching can happen in parallel.
+                    tasks.Add(Task.Run(() => SearchDocumentAsync(document, tempArray)));
+                }
+            }
+
+            private async Task SearchDocumentAsync(
+                Document document, ArrayBuilder<INavigateToSearchResult> results)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
```

Comment:
That's probably true.  But it means that this code needs to know that it is being called directly from Task.Run.  I prefer methods not have to care and to generally follow the rules of:

1. At the entry point check for cancellation.
2. During loops check for cancellation.

This ensures that code doesn't really have to care about interprocedural relationships, and it means we can refactor things and not have to think about what cancellation checks we should or should not have.